### PR TITLE
Leaves - Kegsy (Cloudy, Elizabeth, Kristina, Tiffany)

### PR DESCRIPTION
**Pull request by TiffanyChio**
# bEtsy
Congratulations! You're submitting your assignment! These comprehension questions should be answered by all members of your team, not by a single teammate.

## Comprehension Questions
Question | Answer
:------------- | :-------------
Each team member: what is one thing you were primarily responsible for that you're proud of? | Elizabeth: I worked on the Merchants controller so I am proud of everything in there, but specifically, I am proud of the merchants model methods that connect the merchant to their orderitems. I was able to learn more about querying to get the data I needed. I worked with Kristina on some of the merchant model methods and it was helpful to talk through what messages were being sent where. Kristina: I was primarily responsible for the homepage and general layout of the website, and I was really proud of how the design turned out. Before this project I was uncomfortable with CSS but now I feel like I can take on more front end design in my future projects. Tiffany: I worked on orderitems and I think the logic surrounding adding items to cart was very comprehensive. Adding the same order twice increases the quantity of the existing order item, rather than creating a new one. The code keeps track of orders using the session hash. Cloudy: I started out being responsible for orders and bc its so dependent on so many other factors, I paired with Tiffany who really fleshed it out. I also handled reviews.
Each team member: what is one thing you were primarily responsible for that you would like targeted feedback on? | Elizabeth: I would love to get feedback on drying up the merchant code. I would love to hear where in the views I could have used less logic and moved it into model methods. Kristina: I would like some feedback on my products controller and testing. Tiffany: Orders. Cloudy: Yes pls, reviews.
How did your team break up the work to be done? | We initially built out the ERD and schema together. Then we built out different controllers and came back together to integrate them. We used Trello to keep track of our progress and to communicate what we were working on.
How did your team utilize git to collaborate? | We created a development branch which we worked off of, rather than risk pushing risky code into master. Both development and master branches were protected and required PRs and at least 2 approvers. We also created branches off of the development branch which we used to develop features.
What did your group do to try to keep your code DRY while many people collaborated on it? | One example is our CSS in which we had one member initialize some of the code and classes to provide tools for others to use. We also created many model methods which were easily accessible and reusable.
What was a technical challenge that you faced as a group? | Heroku and logging in with OAuth have been challenging. Many issues popped up for Heroku which didn't for local host. 
What was a team/personal challenge that you faced as a group? | Over the weekend, it was difficult to get PRs in because as a team, we come together to review PRs together while the individual submitting goes over their changes.
What was your application's ERD? (include a link) | (please see your slack messages)
What is your Trello URL? | https://trello.com/betsyproject
What is the Heroku URL of your deployed application? | https://kegsy.herokuapp.com/

### CODE REVIEW

File: app/views/merchants/merchant_orderitems.html.erb [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        <th>Product</th>
+        <th>Date Paid</th>
+        <th>Cost x Quantity</th>
+        <th>Subtotal</th>
+        <th>Status</th>
+        <th>Shipping</th>
+      </tr>
+    </thead>
+    <tbody>
+      <% Merchant.find_merchants_orderitems(@current_merchant, "false").each do |orderitem| %>
```
Comment (beccaelenzil) [inline app/views/merchants/merchant_orderitems.html.erb:27]: re: Elizabeth's request for focused feedback: Could this logic be simplified by making use of the instance variable orderitems, i.e. `@current_merchant.orderitems`

**Comment by beccaelenzil** [comment]
# bEtsy
## What We're Looking For



### Manual testing

Workflow | yes / no
---   | ---
Deployed to Heroku |yes
**Before logging in** |
Browse all products, by category, by merchant |yes
Leave a review |yes
Verify unable to create a new product |yes
**After logging in** | |
Create a category |yes
Create a product in that category with stock 10 |yes
Add the product you created to your cart |yes
Add it again (should update quantity) |yes
Verify unable to increase quantity beyond stock |yes
Add another merchant's product |yes
Check out |yes
Check that stock was reduced |yes
Change order-item's status on dashboard |yes
Verify unable to leave a review for your own product |yes -- I tried an I correctly produced a flash error message. Consider removing this functionality from the UI
Verify unable to edit another merchant's product by manually editing URL |no - I was able to edit ping pong balls
Verify unable to see another merchant's dashboard by manually editing URL |yes, good use of a custom route (`/merchants/current`)

### Code Review

Area | yes / no
---  | ---
**Routes** |
No un-needed routes generated (check `reviews`) | yes
Routes not overly-nested (check products and merchants) |yes
Merchant dashboard and cart page use a non-parameterized routes (should pull merchant or cart ID from session) | yes!
**Controllers** |
Controller-filter to require login by default |yes
Helper methods or filters to find logged-in user, cart, product, etc |yes
No excessive business logic |yes
**Business logic that ought to live in the model** |
Add / remove / update product on order |Good work putting the business in the model
Checkout -> decrease inventory |yes -`reduce_stock`
Merchant's total revenue |yes
Find all orders for this merchant (instance method on `Merchant`) |you implemented this as a class method with the current merchant as an argument. Could this be simplified as an instance method? This would eliminate the work of finding the merchants order items as they could just be accessed as the instance variable `orderitems`.
**Selected Model Tests** |
Add item to cart:<br>  - Can add a good product<br>  - Can't add a product w/o enough stock<br>  - Can't add a retired product<br>  - Can't add to an order that's not in cart mode<br>  - Logic specific to this implementation | yes
Get orders for this merchant:<br>  - Includes all orders from this merchant<br>  - Doesn't include orders from another merchant<br>  - Orders are not included more than once<br>  - Does something reasonable when there are no orders for this merchant | could use a few more tests for edge cases
**Selected Controller Tests** |
Add item to cart:<br>  - Empty cart (should be created)<br>  - Cart already exists (should add to same order)<br>  - Product already in cart (should update quantity)<br>  - Bad product ID, product is retired, quantity too high, or something like that (error) |yes
Leave a review:<br>  - Works when not logged in<br>  - Works when logged in as someone other than the product's merchant<br>  - Doesn't work if logged in as this product's merchant<br>  - Doesn't work if validations fail |yes



## Overall Feedback

Great work overall! You've built a fully functional web store from top to bottom. This represents a huge amount of work, and you should be proud of yourselves!.

I am particularly impressed by how lean and clear your controller actions are, the good use of custom model methods for business logic, and the thoroughness of your tests. 


Requested focused feedback:


Elizabeth: I would love to get feedback on drying up the merchant code. I would love to hear where in the views I could have used less logic and moved it into model methods. 


I left one comment above about changing a class method to an instance method. 


Kristina: I would like some feedback on my products controller and testing. 


Your tests are quite thorough. Good work organizing tests by logged in and guest users. In the logged in user section, you need tests for authorization restrictions with regards to other merchants products. You are missing a test for not being able to test another merchants product. This would have caught this missed requirement. 


The controller is well written. Could you move the logic in toggle retire to the model to make sure all business logic happens in the model?


Tiffany: Orders. 


Great use of custom model methods and controller filters.


Cloudy: Yes pls, reviews.


Clean and thorough tests, model, and controller



bEtsy is a huge project on a very short timeline, and this feedback should not at all diminish the magnitude of what you've accomplished. Keep up the hard work!

# Only the person who submitted the PR will get an email about this feedback. Please let the rest of your team know about it.
